### PR TITLE
KAFKA-17102: FetchRequest#forgottenTopics would return incorrect data

### PR DESCRIPTION
**Pull request by m1a2st**
Jira: https://issues.apache.org/jira/browse/KAFKA-17102
The Cache data will return incorrect data, hence we should remove the cache 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by chia7712 — COMMENTED** [review]
@m1a2st this PR is a kind of bug fix, so please add unit test

**Comment by m1a2st** [comment]
@chia7712, Thanks for your comments, I add test for it, PTAL

File: clients/src/test/java/org/apache/kafka/common/requests/FetchRequestTest.java [hunk 1]
```diff
@@ -243,4 +245,44 @@ public void testPartitionDataEquals() {
...
+        FetchRequest fetchRequest = FetchRequest.Builder
+                .forReplica(version, 0, 1, 1, 1, partitionData)
+                .removed(Collections.emptyList())
+                .replaced(toReplace)
+                .metadata(FetchMetadata.newIncremental(123)).build(version);
+        
+        HashMap<Uuid, String> topicNames = new HashMap<>();
+        topicNames.put(topicId, tp.topic());
+        
+        List<TopicIdPartition> requestsWithTopicsName = fetchRequest.forgottenTopics(topicNames);
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/requests/FetchRequestTest.java:256]: `Collections.singletonMap(topicId, tp.topic())`

**Review by chia7712 — COMMENTED** [review]
@m1a2st nice patch. please take a look at one small question

**Comment by m1a2st** [comment]
@chia7712, Thanks for your comments, PTAL

**Comment by chia7712** [comment]
@m1a2st please rebase code to include your recent fix to trigger QA

**Comment by m1a2st** [comment]
@chia7712, Thanks for reminder, rebased it

File: clients/src/test/java/org/apache/kafka/common/requests/FetchRequestTest.java [hunk 2]
```diff
@@ -243,4 +244,43 @@ public void testPartitionDataEquals() {
             new FetchRequest.PartitionData(Uuid.randomUuid(), 300, 0L, 300, Optional.of(300)));
     }
 
+    @Test
+    public void testFetchRequestNoCacheData() {
+        short version = 13;
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/requests/FetchRequestTest.java]: Do we have test case for version < 13?

Comment (m1a2st) [reply clients/src/test/java/org/apache/kafka/common/requests/FetchRequestTest.java]: Yes, in `testToReplaceWithDifferentVersions` test all version

Comment (chia7712) [reply clients/src/test/java/org/apache/kafka/common/requests/FetchRequestTest.java]: My point was that we should have test to check the `fetchRequest.forgottenTopics` for all versions. `testToReplaceWithDifferentVersions` does NOT do that

Comment (m1a2st) [reply clients/src/test/java/org/apache/kafka/common/requests/FetchRequestTest.java]: I test all version for `testFetchRequestNoCacheData`, PTAL

File: clients/src/test/java/org/apache/kafka/common/requests/FetchRequestTest.java [hunk 3]
```diff
@@ -243,4 +244,45 @@ public void testPartitionDataEquals() {
...
+        int partition = 0;
+        TopicIdPartition tp = new TopicIdPartition(topicId, partition, "topic");
+
+        Map<TopicPartition, FetchRequest.PartitionData> partitionData = Collections.singletonMap(tp.topicPartition(),
+                new FetchRequest.PartitionData(topicId, 0, 0, 0, Optional.empty()));
+        List<TopicIdPartition> toReplace = Collections.singletonList(tp);
+
+        FetchRequest fetchRequest = FetchRequest.Builder
+                .forReplica(version, 0, 1, 1, 1, partitionData)
+                .removed(Collections.emptyList())
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/requests/FetchRequestTest.java]: Could you add some "removed" TopicIdPartition? That can avoid empty `forgottenTopicMap`

**Review by chia7712 — COMMENTED** [review]
@m1a2st thanks for updated PR!

File: clients/src/test/java/org/apache/kafka/common/requests/FetchRequestTest.java (hunk 3 above)
Comment (m1a2st) [reply clients/src/test/java/org/apache/kafka/common/requests/FetchRequestTest.java]: I according to diffenet version to generate fetchRequest, that can avoid empty `forgottenTopicMap`, PTAL

**Pull request merged**